## Algorithms by Stanford University - Coursera Specialization

## Course 1 - Divide and Conquer, Sorting and Searching, and Randomized Algorithms

## Assignment 1 - Karatsuba Multiplication

In [ ]:
def integer_multiply(x, y):
  result = 0
  y = str(y)
  for i in range(len(y)-1, -1, -1):
    # print(i)
    m = int(y[i])*x
    m = int(str(m) + (len(y) -1 - i)*str(0))
    result += m

  return result

integer_multiply(3141592653589793238462643383279502884197169399375105820974944592, 2718281828459045235360287471352662497757247093699959574966967627)

8539734222673567065463550869546574495034888535765114961879601127067743044893204848617875072216249073013374895871952806582723184

In [ ]:
def karatsuba_multiply(x, y):
    x, y = str(x), str(y)

    if len(x) == 1 or len(y) == 1:
        return int(x) * int(y)
    m = len(x) // 2
    a, b = int(x[:-m]), int(x[-m:])
    c, d = int(y[:-m]), int(y[-m:])

    ac = karatsuba_multiply(a, c)
    bd = karatsuba_multiply(b, d)
    ad_plus_bc = karatsuba_multiply(a + b, c + d) - ac - bd

    result = integer_multiply((10**(2 * m)), ac) + integer_multiply((10**m), ad_plus_bc) + bd

    return result

karatsuba_multiply(3141592653589793238462643383279502884197169399375105820974944592, 2718281828459045235360287471352662497757247093699959574966967627)

8539734222673567065463550869546574495034888535765114961879601127067743044893204848617875072216249073013374895871952806582723184

## Assignment 2 - Counting Inversions with Merge Sort

Merge Sort recursively with Counting Inversions

In [ ]:
import numpy as np

def merge(left, right):
  count = 0
  result = []
  i, j = 0, 0

  while i < len(left) and j < len(right):
      if left[i] <= right[j]:
        result.append(left[i])
        i+= 1

      else:
        count += len(left) - i
        result.append(right[j])
        j+= 1

  result += left[i:]
  result += right[j:]

  return result, count

In [ ]:
def merge_sort(array):
  if len(array) <= 1:
    return array, 0

  mid = len(array)//2
  left, count1 = merge_sort(array[:mid])
  right, count2 = merge_sort(array[mid:])
  result, count3 = merge(left, right)

  total_count = count1 + count2 + count3

  return result, total_count

In [ ]:
from google.colab import files
uploaded = files.upload()
with open("count.txt", "r") as f:
    arr = [int(line.strip()) for line in f]

Saving count.txt to count.txt


In [ ]:
inversions = merge_sort(arr)[1]
inversions

2407905288

## Assignment 3 - QuickSort using Partitioning

In [ ]:
def Partition(A, l, r): ## here pivot is the 1st element

  p = A[l] # pivot elem
  i = l+1
  comp = 0
  for j in range(l+1, r+1):

    if A[j] < p:

      A[j], A[i] = A[i], A[j]

      i+=1
  A[l], A[i-1] = A[i-1], A[l]
  return i-1

# def Partition(A, l, r): this was another choice of pivot
#     A[l], A[r] = A[r], A[l]  # Move pivot to front
#     p = A[l]
#     i = l + 1
#     for j in range(l + 1, r + 1):
#         if A[j] < p:
#             A[j], A[i] = A[i], A[j]
#             i += 1
#     A[l], A[i - 1] = A[i - 1], A[l]
#     return i - 1


In [ ]:
def QuickSort(A, l, r):
    if l >= r:
        return 0

    comparisons = r - l  #Number of comparisons at this level

    p = Partition(A, l, r)

    left_comparisons = QuickSort(A, l, p - 1)
    right_comparisons = QuickSort(A, p + 1, r)

    return comparisons + left_comparisons + right_comparisons

In [ ]:
from google.colab import files
uploaded = files.upload()
with open("quicksort.txt", "r") as f:
    arr = [int(line.strip()) for line in f]

Saving quicksort.txt to quicksort (6).txt


In [ ]:
comparison = QuickSort(arr, 0, len(arr)-1)
print(comparison)

164123


Results:
- Nb of Comparisons when pivot is 1st element: 162085
- Nb of Comparisons when pivot is last element: 164123
- Nb of Comparisons when pivot is median: 138382


## Assignment 4 - Minimum Cut

!! For this part I referred to online resources.

In [ ]:
import random
import copy

# Load the adjacency list from a file
def load_graph(filename):
    graph = {}
    with open(filename, 'r') as f:
        for line in f:
            parts = list(map(int, line.strip().split()))
            node = parts[0]
            graph[node] = parts[1:]
    return graph

# Random contraction algorithm
def karger_min_cut(graph):
    graph = copy.deepcopy(graph)

    while len(graph) > 2:
        # Choose a random edge
        u = random.choice(list(graph.keys()))
        v = random.choice(graph[u])

        # Merge v into u
        graph[u].extend(graph[v])

        for node in graph[v]:
            graph[node] = [u if x == v else x for x in graph[node]]

        # Remove self-loops
        graph[u] = [x for x in graph[u] if x != u]

        # Delete v
        del graph[v]

    # Return the number of crossing edges (edges between the two remaining supernodes)
    return len(list(graph.values())[0])

# Repeat many times to find the min cut
def repeated_karger(filename, trials=200):
    min_cut = float('inf')
    original_graph = load_graph(filename)
    for _ in range(trials):
        cut = karger_min_cut(original_graph)
        min_cut = min(min_cut, cut)
    return min_cut

# Example usage
filename = '/content/min cut.txt'  # Replace with your actual file path
print("Minimum cut found:", repeated_karger(filename, trials=300))


Minimum cut found: 17


## Course 2 - Graph Search, Shortest Paths and Data Structures

## Assignment 1 - Strongly Connected Components using Depth First Search

In [ ]:
def load_graph(filename):
    graph = {i: [] for i in range(1, 875715)}
    with open(filename, 'r') as f:
        for line in f:
            parts = list(map(int, line.strip().split()))
            node = parts[0]
            graph[node].append(parts[1])
    return graph


def load_graph_reversed(filename):

  graph_rev = {i: [] for i in range(1, 875715)}
  with open(filename, 'r') as f:
      for line in f:
          parts = list(map(int, line.strip().split()))
          node = parts[1]
          graph_rev[node].append(parts[0])
  return graph_rev


filename= '/content/scc.txt'
graph = load_graph(filename)
graph_rev = load_graph_reversed(filename)

In [ ]:
import sys
sys.setrecursionlimit(10**6)  # increase recursion depth

## Depth First Search
def DFS(graph, s, visited, stack):

  visited.add(s)
  for v in graph[s]:
    if v not in visited:
      DFS(graph, v, visited, stack)

  stack.append(s)


def DFS_loop(graph, s, visited, current_SCC): # to compute the scc at the end

  visited.add(s)
  current_SCC.append(s)
  for v in graph[s]:
    if v not in visited:
      DFS_loop(graph, v, visited, current_SCC)


def compute_SCC(graph, graph_rev):

  #1 we need to pass on original graph
  visited= set()
  stack= []

  for node in graph:
    if node not in visited:
      DFS(graph, node, visited, stack)

  #2 we need to pass on reversed graph

  visited = set()
  SCCs = []

  while stack:
    node= stack.pop()
    if node not in visited:
      current_SCC= []
      DFS_loop(graph_rev, node, visited, current_SCC)
      SCCs.append(current_SCC)

  return SCCs

In [ ]:
SCCs = compute_SCC(graph, graph_rev)
SCCs.sort(key=len, reverse=True)
SCCs[:5]

[[1,
  5,
  7,
  6,
  133568,
  115322,
  115304,
  8791,
  8775,
  8772,
  8778,
  8773,
  8779,
  8780,
  8781,
  8776,
  8794,
  8818,
  8817,
  23143,
  3123,
  3101,
  3092,
  3094,
  3095,
  3096,
  3097,
  3102,
  3098,
  3099,
  3100,
  3103,
  408961,
  408962,
  408964,
  408965,
  408966,
  408967,
  408968,
  3104,
  408963,
  3105,
  3106,
  3111,
  53383,
  53367,
  53362,
  53363,
  53364,
  53366,
  53368,
  53370,
  53387,
  68431,
  53386,
  53369,
  53380,
  53381,
  53384,
  53385,
  67916,
  53371,
  53388,
  67904,
  67883,
  30566,
  30538,
  30523,
  30526,
  30524,
  30525,
  30527,
  30528,
  30530,
  30534,
  30535,
  30536,
  30539,
  30540,
  30541,
  30542,
  30543,
  30544,
  30545,
  30546,
  30547,
  30548,
  30549,
  30550,
  30551,
  30552,
  30565,
  115370,
  115308,
  8806,
  8821,
  3112,
  8782,
  8777,
  8799,
  8796,
  104388,
  8792,
  8793,
  8795,
  8813,
  8816,
  8823,
  8784,
  8814,
  100331,
  100296,
  100288,
  100289,
  100290,
  100

In [ ]:
print(len(SCCs[0]), len(SCCs[1]), len(SCCs[2]), len(SCCs[3]), len(SCCs[4]))

434821 968 459 313 211


## Assignment 2 - Dijkstra's Shortest Path Algorithm

In [ ]:
import numpy as np
import pandas as pd
import heapq
from google.colab import files

In [ ]:
def read_graph(filename):

  graph={i: [] for i in range(1, 201)}

  with open(filename, 'r') as f:
    for line in f:

      parts=line.strip().split() # now we have smhtg like ['6', '3,5', '4,9'...]
      if not parts: continue

      u=int(parts[0]) # the node
      for pair in parts[1:]: # the list of edges
        v, w = pair.split(',')
        v,w = int(v), int(w) ## now we have node u, with edge to v with the weight w

        graph[u].append((v, w)) # add this to the adjacecny list
        graph[v].append((u, w)) # this is bcz we have an undirected graph
    return graph

In [ ]:
def dijkstra(graph, source):

    ## now we initialize the distance graph with +inf
    distance = {i: float('inf') for i in graph.keys()}
    distance[source] = 0 ## we initialize the first distance to 0

    visited = {i: False for i in graph.keys()}

    heap = [(0, source)] # this stores the following (distance, vertex)

    while heap:

      dist,u = heapq.heappop(heap) ## this extracts vertex u with smallest known distance
      if visited[u]: continue ## skip

      visited[u] = True

      for v, w in graph[u]:

        if not visited[v]:
          alt= dist + w ## this is the distance to go to node v from edge u (prev dist + weight)
          if alt < distance[v]:
            distance[v]=alt ## this the shortest path to v till now
            heapq.heappush(heap, (alt, v))

    ## update the un-updated dist
    for v in graph.keys():
      if distance[v] == float('inf'):
        distance[v]= 1000000

    return distance

In [ ]:
graph = read_graph('/content/dijkstraData.txt')

In [ ]:
distance = dijkstra(graph, 1)
for destination in [7,37,59,82,99,115,133,165,188,197]:
  print(distance[destination])

2599
2610
2947
2052
2367
2399
2029
2442
2505
3068


## Assignment 3 - Median Maintenance using Heaps

In [ ]:
import numpy as np
import pandas as pd
import heapq
from google.colab import files

uploaded = files.upload()
with open("heaps.txt", "r") as f:
    numbers = [int(line.strip()) for line in f]

Saving heaps.txt to heaps.txt


In [ ]:
lower_heap= [-numbers[0]]
upper_heap= []
median = [numbers[0]]

In [ ]:
for number in numbers[1:]:

  if number <= - lower_heap[0]:
    heapq.heappush(lower_heap, -number)
  else:
    heapq.heappush(upper_heap, number)

  if len(lower_heap) > len(upper_heap)+1:
    n = - heapq.heappop(lower_heap)
    heapq.heappush(upper_heap, n)

  elif len(upper_heap) > len(lower_heap)+1:
    n = heapq.heappop(upper_heap)
    heapq.heappush(lower_heap, -n)

  if len(lower_heap) == len(upper_heap):
    median.append(-lower_heap[0])
  else:
    if len(lower_heap) > len(upper_heap):
      median.append(-lower_heap[0])
    else:
      median.append(upper_heap[0])

In [ ]:
# compute the sum of median modulo 1000
sum_median = sum(median) % 10000
print(sum_median)

1213


Final assignment is not available